### PR TITLE
Elevate Linux Permissions,Introduce Command Line Switches

### PR DESCRIPTION
**Pull request by josegomez**
This Pull Request Addresses issue #196 
It also introduces command line switch support in the Installer for actions
Install, Uninstall and (eventual) Update #197 (Separate PR)

I have elevation working pretty darn well see below 
Elevation

Behaves just like UAC

I'm currently fighting with a different issue, when you elevate then your "current" user is root and all your special folders point to /root/.config 

So the install PathConf ends up in the wrong location. I am trying to find a way how to figure out who the current user is while the process is running as root.

My current (and only) solution is to pass in the current user Home directory when I elevate... I'm still investigating.

The way the elevation works is as follows. Once you have installed SQRL the first time if [polkit ](https://www.freedesktop.org/software/polkit/docs/latest/polkit.8.html) is available
We create an execution rile for SQRLPlatformInstaller_linux that basically says this program is allowed to run as admin.

When the installer launches (as a regular user) it checks to see if its sudo/root. If it is not it asks polkit to re-launch itself with elevated permissions.

User is prompted and if approved program runs (now as root). Here is where the issue comes into play at this point all the Special Folders and Current Users see "root" (not the original user)

I have thus far been unable to figure out a way around this, my current thought is that when the program runs at first (as current user) right before we ask polkit for an elevation we pass in a command line switch -c | -currentuser which we can then use to map the paths.

This would require a special exception in PathConf for Linux but I don't see another solution.

We require root to copy files into /opt/SQRL and we require root to generate the pokit action / rules for elevation so root is a requirement if we want this to work, but as root we lose the current user....

Don't you just LOVE linux? 😫🤐

PS: Here's an example of what I'm talking about I was able to run the elevation from the console and capture the log output, note the Config Path (even though i'm user josecgomez)
RootPath

### CODE REVIEW

**Comment by josegomez** [comment]
@alexhauser I threw the review request out there while I'm investigating in case you have any feedback or ideas in the mean time.

**Comment by alexhauser** [comment]
@josegomez, first of all, this is looking great! 

I would love it if we could get the path mapping to work with this, as the elevation process looks as user-friendly as it can be. 

Also, your approach on the command line parsing is beautiful, I love it!

As for the path mapping issue: 
This is really infuriating. `sudo` does handle this very well. Even in a `sudo` session, `$HOME` will map to the original user's home path, which is what you want almost all of the time. The more disappointing it is, that it seems PolicyKit doesn't support this behaviour.

If we try to map the paths locally, you could try using the `logname` command instead of having to pass around command line arguments, since that should give you the name of the currently logged in user, irrespective of any "elevation" level. I haven't tried it with PolicyKit, but it should hopefully work there as well.

I will give this some more thought as well, and poke around a little.

Proper code review will follow once the PR is complete.

**Comment by josegomez** [comment]
Yeah polkit does more of a run as than a sudo invoke unfortunately 
I’ll keep working the problem

**Comment by josegomez** [comment]
I just had an idea not sure how well it will work but if we run export HOME = path/to/current/user/home 
That should remap HOME to the right place for the session 
I wonder if then the special folder stuff will work correctly...
I gotta assume they use environment variables to determine

**Comment by alexhauser** [comment]
@josegomez, have you tried `logname` as suggested above?

```C#
string user = _shell.Term("logname", Output.Hidden).stdout.Trim();
```

**Comment by josegomez** [comment]
@alexhauser I just checked and logname does show the current logged on user correctly.!

**Comment by alexhauser** [comment]
Cool, then we do at least have another option, @josegomez.

You should be able to get the full home path of the user like this:

```C#
string user = _shell.Term("logname", Output.Hidden).stdout.Trim();
string home = _shell.Term($"getent passwd {user} | cut -d: -f6", Output.Hidden).stdout.Trim();
```

This is from here: https://unix.stackexchange.com/questions/247576/how-to-get-home-given-user

If we add that to `PathConf` if we detect that `Util.IsAdmin() == true`, we should be good, no?

**Comment by josegomez** [comment]
yeah that should work .net core uses the following to get the special folders
image

And this to get the Home Directory
image

**Comment by alexhauser** [comment]
Ah, they fall back to `/etc/passwd` too if the environment variable does not exist. The `getent` tool also looks into `/etc/passwd`, since that's the only place where the home directory is guaranteed to exist.

**Comment by josegomez** [comment]
Well what a CLUSTER ! I think I got it... If you don't mind @alexhauser when looking through the code be a bit more thorough than normal there were a lot of things I had to change and do this will require some serious testing before release.

Current Experience should be as follows 
Initial Install:
Because upon Initial Inistall we won't be able to automatically elevate (no polikit rules in place) the software will prompt you to run as sudo
InitialInstall


Updates and Subsequent Installs:
Updates and subsequent Installs will (should) automatically Elevate and execute.
Updates

The logic for this got pretty complex, the initial install creates the polkit rules so that the file can be elevated. These polkit rules require a specific Path so I make this path be 
"/tmp/SQRLPlatformAwareInstaller_linux"

The reason for this is as follows

When the installer runs (as non root) it will automatically copy itself into /tmp/ so that when the elevation occurs the latest (current) copy of the Installer exists in /tmp/ 

The reason I put it in /tmp/ is because if we ran it directly from /opt/SQRL then when  the unzip operation occurs the installer is in use and causes conflict.

So to summarize
If the installer runs as non root the first time
It asks the user to run as sudo and bails
Then the install process runs as normal and it creates the polkit rules

Next Time
If the Installer runs as non root

- The installer copies itself to /tmp/SQRLPlatformAwareInstaller_linux
- The installer creates a script in temp which invokes polkit and the above copy of the installer
- The installer bails
- The above script invokes polkit elevation and launches /tmp/SQRLPlatformAwareInstaller_linux (as root)
- Because the program is running from /tmp/SQRLPlatformAwareInstaller_linux there is no conflict with over-writing or copying files into /opt/SQRL where it originally lives

File: SQRLCommonUI/Models/PathConf.cs [hunk 1]
```diff
@@ -24,9 +28,26 @@ public static class PathConf
...
+            {
+                string cfPath = "";
+                /*
+                 * If this is running on linux we and running as admin we need to hack our way
+                 * to find the current user's home directory so that we get the correct config path
+                 */
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) && AdminCheck.IsAdmin())
+                {
+                    string user = _shell.Term("logname", Output.Hidden).stdout.Trim();
+                    string home = _shell.Term($"getent passwd {user} | cut -d: -f6", Output.Hidden).stdout.Trim();
```
Comment (alexhauser) [inline SQRLCommonUI/Models/PathConf.cs]: To avoid duplicating code, I would suggest pulling this out into a function á la 
```C#
public static string GetHomePath()
```

File: SQRLCommonUI/Models/PathConf.cs [hunk 2]
```diff
@@ -40,8 +61,29 @@ public static class PathConf
...
+            {
+                string defaultDbPath = "";
+                /*
+                * If this is running on linux we and running as admin we need to hack our way
+                * to find the current user's home directory so that we get the correct Default DB path
+                */
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) && AdminCheck.IsAdmin())
+                {
+                    string user = _shell.Term("logname", Output.Hidden).stdout.Trim();
+                    string home = _shell.Term($"getent passwd {user} | cut -d: -f6", Output.Hidden).stdout.Trim();
```
Comment (alexhauser) [inline SQRLCommonUI/Models/PathConf.cs]: See my comment about pulling this out into its own method.

File: SQRLCommonUI/Models/PathConf.cs [hunk 3]
```diff
@@ -103,8 +145,10 @@ public static bool ConfigFileExists
         /// </summary>
         public static void LoadConfig()
         {
+            Log.Information($"Config File:{ConfFile}");
```
Comment (alexhauser) [inline SQRLCommonUI/Models/PathConf.cs]: Missing space after colon. (You told me to be thorough 😉)

File: SQRLCommonUI/SQRLCommonUI.csproj [hunk 4]
```diff
@@ -20,7 +28,17 @@
...
     <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Update="C:\Users\jose\.nuget\packages\dein.toolbox\1.6.4\contentFiles\any\any\cmd.bat">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="C:\Users\jose\.nuget\packages\dein.toolbox\1.6.4\contentFiles\any\any\cmd.sh">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
```
Comment (alexhauser) [inline SQRLCommonUI/SQRLCommonUI.csproj]: Hmmm... full local paths, that doesn't look right. Or is it?

File: SQRLPlatformAwareInstaller/App.xaml.cs [hunk 5]
```diff
@@ -1,24 +1,95 @@
...
+                            if(string.IsNullOrEmpty(result.stderr.Trim()) && File.Exists(Path.Combine("/usr/share/polkit-1/actions", "org.freedesktop.policykit.SQRLPlatformAwareInstaller_linux.policy")))
+                            {
+                                Log.Information("Found Installer in Path!");
+                                
+                                /*Polkit invokation forbids having a "dead" parent, so if we invoke polkit directly from here
+                                  and then kill the process it will abort. First we need to write the polkit invocation
+                                  to a shell script which is invoked externally so that we can kill our current instance of the installer cleanly.
+                                */
+                                var tmpScript = Path.GetTempFileName().Replace(".tmp",",sh");
+                                Log.Information($"Copying Installer From:{Process.GetCurrentProcess().MainModule.FileName} to: /tmp/SQRLPlatformAwareInstaller_linux");
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/App.xaml.cs]: Missing space after colon.

File: SQRLPlatformAwareInstaller/Assets/Localization/localization.json [hunk 6]
```diff
@@ -98,7 +105,10 @@
       "NoReleasesMessage": "Keine Releases verfügbar? Probieren Sie, Vorab-Releases zu aktivieren!",
       "TitleUninstall": "SQRL Client deinstallieren",
       "UninstallMessage": "Sie sind im Begriff, die SQRL Client-App von Ihrem Computer zu entfernen.\r\n\r\nSämtliche Programm-Daten werden dabei gelöscht. Ihre Benutzerdaten (z.B. ihre SQRL-Identitäten) bleiben dabei erhalten. Drücken Sie \"Deinstallieren\", um fortzufahren.",
-      "BtnUninstall": "Deinstallieren"
+      "BtnUninstall": "Deinstallieren",
+      "BtnExit": "Ausfahrt",
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Assets/Localization/localization.json]: LOL, this would be the "highway exit". Correct translation is "Beenden".

File: SQRLPlatformAwareInstaller/Assets/Localization/localization.json [hunk 7]
```diff
@@ -98,7 +105,10 @@
...
+      "BailTitle": "FEHLER: Das SQRL-Installationsprogramm erfordert Root-Zugriff",
+      "RootAccessRequired": "Für die Installation / Aktualisierung von SQRL ist Root- oder Administratorzugriff erforderlich.\r\n\r\nDie Sitzung konnte nicht automatisch erhöht werden. Schließen Sie das Installationsprogramm und führen Sie es als sudo aus, um es erneut zu versuchen"
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Assets/Localization/localization.json]: That's almost right! 😁

`Für die Installation / Aktualisierung von SQRL ist Root- oder Administratorzugriff erforderlich.\r\n\r\nDie Berechtigung der aktuellen Sitzung konnte nicht automatisch erhöht werden. Schließen Sie das Installationsprogramm und führen Sie es mit \"sudo\" erneut aus.`

File: SQRLPlatformAwareInstaller/Assets/SQRLPlatformAwareInstaller_linux.policy [hunk 8]
```diff
@@ -0,0 +1,20 @@
+﻿<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE policyconfig PUBLIC
+ "-//freedesktop//DTD PolicyKit Policy Configuration 1.0//EN"
+ "http://www.freedesktop.org/standards/PolicyKit/1/policyconfig.dtd">
+<policyconfig>
+  <action id="org.freedesktop.policykit.pkexec.SQRLPlatformAwareInstaller_linux">
+    <description>Run SQRL Platform Installer  Program</description>
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Assets/SQRLPlatformAwareInstaller_linux.policy]: Double space

File: SQRLPlatformAwareInstaller/Models/CommandLineArgs.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using CommandLine;
+
+namespace SQRLPlatformAwareInstaller.Models
+{
+    /// <summary>
+    /// This enum is used by the installer to determine the action that was requested to be performed
+    /// by default it would be Install, but we also have the option to Unininstall and eventually upgrade the application.
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Models/CommandLineArgs.cs]: typo in "Uninstall"

File: SQRLPlatformAwareInstaller/Models/CommandLineArgs.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    public class CommandLineArgs
+    {
+        /// <summary>
+        /// Command line switch -a | -action <see cref="InstallerAction"/>
+        /// </summary>
+        [Option('a',"action",Required =false, Default = InstallerAction.Install, HelpText = "Specifies the action to be performed (Install, Uninstall, Update).")]
+        public InstallerAction Action { get; set; }
+
+        /// <summary>
+        /// Command line switch -p | -path "path hold's SQRL current install path
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Models/CommandLineArgs.cs]: Description you used below is better: `"Specifies SQRL's current Install path (directory)"`

File: SQRLPlatformAwareInstaller/Models/CommandLineArgs.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        /// </summary>
+        [Option('p', "path", Required = false,  Default ="", HelpText = "Specifies SQRL's current Install path (directory)")]
+        public string InstallPath { get; set; }
+
+        /// <summary>
+        /// Command line switch -z | zip "path to update zip" holds the path ofthe downloaded update zip
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Models/CommandLineArgs.cs]: ofthe (missing space).
Also, the description you use below is nicer: `"Specifies the location of the temporary update zip file (used in conjunction with update action)"`

File: SQRLPlatformAwareInstaller/Platform/Linux/Installer.cs [hunk 12]
```diff
@@ -63,19 +65,74 @@ public async Task Install(string archiveFilePath, string installPath, string ver
...
+                string user = _shell.Term("logname", Output.Hidden).stdout.Trim();
+                string home = _shell.Term($"getent passwd {user} | cut -d: -f6", Output.Hidden).stdout.Trim();
+                if (AdminCheck.IsAdmin())
+                {
+                    _shell.Term($"update-desktop-database {home}/.local/share/applications/", Output.Internal);
+                }
+                else
+                {
+                    _shell.Term($"update-desktop-database ~/.local/share/applications/", Output.Internal);
+                }
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Platform/Linux/Installer.cs]: If we implement `GetHomePath()` in `PathConf` as suggested, we could even make it `public static` and use it here as well. Ideally, this whole upper block could then be replaced with:

```C#
var home = PathConf.GetHomePath();
_shell.Term($"update-desktop-database {home}/.local/share/applications/", Output.Internal);
```

Should this `getent`-hack bite us in the future, it would just be so much cleaner to only have to change it in one central place instead of needing to search every instance of it being used scattered around the project.

File: SQRLPlatformAwareInstaller/Platform/Linux/Installer.cs [hunk 13]
```diff
@@ -63,19 +65,74 @@ public async Task Install(string archiveFilePath, string installPath, string ver
...
 
                 // Change owner of database dir/file to the actual user behind the "sudo"
-                string user = _shell.Term("logname", Output.Hidden).stdout.Trim();
+                
                 string chownDbFile = $"chown -R {user}:{user} {PathConf.ClientDBPath}";
                 Log.Information($"Determined username for chown: \"{user}\"");
                 Log.Information($"Running command: {chownDbFile}");
                 _shell.Term(chownDbFile, Output.Internal);
 
+                Log.Information("All us good up to this point, lets setup Linux for UAC (if we can)");
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Platform/Linux/Installer.cs]: "All **i**s good..." (typo)

File: SQRLPlatformAwareInstaller/Platform/Linux/Installer.cs [hunk 14]
```diff
@@ -63,19 +65,74 @@ public async Task Install(string archiveFilePath, string installPath, string ver
...
+
+                /*
+                 * Creates the required file and system changes for SQRL to be available
+                 * ubiquitous throughout the system via a new 
+                 * environment variable SQRL_HOME and the addition of this variable to the system PATH.
+                 * 
+                 * Note that the later won't take effect until the user logs out or reboots
+                 */
+                var result = _shell.Term("command -v pkexec", Output.Internal);
+                if (string.IsNullOrEmpty(result.stderr) && !string.IsNullOrEmpty(result.stdout))
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Platform/Linux/Installer.cs]: This check is also used more than once throughout the code.
Something like `AdminCheck.IsPolicyKitInstalled()` or similar could make everything easier to read and maintain.

File: SQRLPlatformAwareInstaller/Platform/Linux/Installer.cs [hunk 15]
```diff
@@ -84,7 +141,17 @@ public async Task Uninstall(IProgress<Tuple<int, string>> progress = null, bool
...
+            string home = _shell.Term($"getent passwd {user} | cut -d: -f6", Output.Hidden).stdout.Trim();
+            if (AdminCheck.IsAdmin())
+            {
+                _shell.Term($"update-desktop-database {home}/.local/share/applications/", Output.Internal);
+            }
+            else
+            {
+                _shell.Term($"update-desktop-database ~/.local/share/applications/", Output.Internal);
+            }
+
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Platform/Linux/Installer.cs]: Same here, this conditional shouldn't be necessary when using `PathConf.GetHomePath()`.

File: SQRLPlatformAwareInstaller/Utils.cs [hunk 16]
```diff
@@ -137,6 +115,10 @@ public static void GrantFullFileAccess(string file)
                 fi.SetAccessControl(ac);
 
             }
+            else if ((RuntimeInformation.IsOSPlatform(OSPlatform.Linux)))
+            {
+                _shell.Term($"chmod 755 {file}", Output.Hidden);
+            }
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Utils.cs]: Any special reason for adding this? Shouldn't our call to

```c#
_shell.Term($"chmod -R 755 {installPath}", Output.Internal);
```
after the extraction already take care of this?

File: SQRLPlatformAwareInstaller/ViewModels/MainWindowViewModel.cs [hunk 17]
```diff
@@ -35,18 +36,23 @@ public ViewModelBase PriorContent
         /// <summary>
         /// Creates a new instance and sets the content of the window.
         /// </summary>
-        public MainWindowViewModel()
+        public MainWindowViewModel(bool rootBail=false)
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/ViewModels/MainWindowViewModel.cs]: Some documentation for the new parameter would be nice here.

File: SQRLPlatformAwareInstaller/ViewModels/MainWindowViewModel.cs [hunk 18]
```diff
@@ -35,18 +36,23 @@ public ViewModelBase PriorContent
...
             {
-                Log.Information($"Installer was called with \"{args[1]}\" command line switch - launching uninstall screen");
-                viewModel = new UninstallViewModel();
+                viewModel = new RootBailViewModel();
             }
             else
+            if (InstallerCommands.Instance != null && InstallerCommands.Instance.Action == InstallerAction.Uninstall)
+            {
+
+
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/ViewModels/MainWindowViewModel.cs]: Excess whitespace.

File: SQRLPlatformAwareInstaller/ViewModels/MainWindowViewModel.cs [hunk 19]
```diff
@@ -35,18 +36,23 @@ public ViewModelBase PriorContent
...
+                Log.Information($"Installer was called with \"{InstallerCommands.Instance}\" command line switches - launching uninstall screen");
+                viewModel = new UninstallViewModel();
+            }
+            else if(InstallerCommands.Instance.Action== InstallerAction.Install)
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/ViewModels/MainWindowViewModel.cs]: Above, we do a `null` check on `InstallerCommands.Instance`.
Is it not required here as well?

File: SQRLCommonUI/Models/PathConf.cs [hunk 20]
```diff
@@ -40,8 +61,29 @@ public static class PathConf
...
+                    string home = _shell.Term($"getent passwd {user} | cut -d: -f6", Output.Hidden).stdout.Trim();
+                    defaultDbPath = Path.Combine(home, "SQRL");
+                }
+                else
+                    defaultDbPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), "SQRL");
+
+                return defaultDbPath;
+            }
+
+
```
Comment (alexhauser) [inline SQRLCommonUI/Models/PathConf.cs]: Excess whitespace.

File: SQRLPlatformAwareInstaller/App.xaml.cs [hunk 21]
```diff
@@ -1,24 +1,95 @@
...
+                                _shell.Term($"chmod a+x {tmpScript}");
+                                Process proc = new Process();
+                                proc.StartInfo.FileName = tmpScript;
+                                proc.Start();
+                                nogo=false;
+                            }
+                            
+                        }
+                        
+
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/App.xaml.cs]: Excess whitespace.

File: SQRLPlatformAwareInstaller/Views/RootBailView.xaml [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+         />
+    
+
+   
+
+    <TextBlock TextAlignment="Center" FontWeight="Bold" MaxWidth="580" Grid.Row="1" Grid.ColumnSpan="3" Grid.Column="0" Text="{loc:Localization RootAccessRequired}"  Margin="10" HorizontalAlignment="Center" Foreground="Black" FontSize="15" TextWrapping="Wrap"  />
+
+
+    
+
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/Views/RootBailView.xaml]: Whitespace/formatting.

**Review by alexhauser — CHANGES_REQUESTED** [review]
Fantastic job Jose, great to see that you got everything working! 🎉

I guess you'll probably regret asking me to be thorough. But never the less, here you go with my (somewhat pedantic) review.

In general, we could maybe look into encapsulating all the PolicyKit-related nitty-gritty into its own class (or maybe use the `AdminCheck` class), in order to de-clutter the higher-level "logic"-code in the Installer and in App classes. Don't know if it's possible, but could be worth exploring.

**Comment by josegomez** [comment]
Haha no problem this was a big and troublesome change better thorough than not! 
Thanks !

File: SQRLCommonUI/Models/PathConf.cs (hunk 3 above)
Comment (josegomez) [reply SQRLCommonUI/Models/PathConf.cs]: Bahaha 😂 fair enough

File: SQRLCommonUI/SQRLCommonUI.csproj (hunk 4 above)
Comment (josegomez) [reply SQRLCommonUI/SQRLCommonUI.csproj]: It did that on its own since these files come from Nuggets I’m not sure how to fix it I’ll poke about

File: SQRLPlatformAwareInstaller/Assets/Localization/localization.json (hunk 6 above)
Comment (josegomez) [reply SQRLPlatformAwareInstaller/Assets/Localization/localization.json]: lol I’ll blame google for this

File: SQRLPlatformAwareInstaller/Utils.cs (hunk 16 above)
Comment (josegomez) [reply SQRLPlatformAwareInstaller/Utils.cs]: I missed that yup!

File: SQRLCommonUI/Models/PathConf.cs (hunk 1 above)
Comment (josegomez) [reply SQRLCommonUI/Models/PathConf.cs]: Added 2 Methods to AdminCheck GetUser and GetHome

File: SQRLCommonUI/Models/PathConf.cs (hunk 3 above)
Comment (josegomez) [reply SQRLCommonUI/Models/PathConf.cs]: Fixed

File: SQRLPlatformAwareInstaller/App.xaml.cs [hunk 23]
```diff
@@ -1,24 +1,92 @@
...
+                        Log.Information("Launched on Linux without Sudo, trying to re-launch if possible");
+                        Log.Information("Checking if pkexec exists"); //This allows us to elevate a program
+                        
+                        //Checks to see if pkexec exists, if it doesn't bail #NothingWeCanDo
+                        var result = _shell.Term("command -v pkexec", Output.Internal);
+                        if(string.IsNullOrEmpty(result.stderr.Trim()) && !string.IsNullOrEmpty(result.stdout.Trim()))
+                        {
+                            //Checks to make sure that SQRL_HOME environment variable exists as well as the SQRLPlatformInstaller this is needed for the polkit invokation
+                            string pkexec = result.stdout.Trim();
+                            Log.Information("pkexec exists!");
```
Comment (alexhauser) [inline SQRLPlatformAwareInstaller/App.xaml.cs]: Why not  use `SystemAndShellUtils.IsPolKitAvailable()` here?

**Review by alexhauser — APPROVED** [review]
Looks good Jose, only one more minor comment.
Merge away!

**Comment by josegomez** [comment]
@alexhauser made a few more modifications.

**Comment by alexhauser** [comment]
LGTM! 👍🏻

**Review by alexhauser — APPROVED** [review]
